air-quality-backend/src/air_quality/etl/forecast/forecast_adapter.py

```diff
@@ -107,3 +113,309 @@ def transform(forecast_data: ForecastData, locations: list[AirQualityLocation])
             }
             pollutant_forecast_for_location.append(document)
     return pollutant_forecast_for_location
+
+
+def _get_dimension_by_attr(
+    dataset: xr.Dataset, attr_name: str, attr_value: str
+) -> Optional[xr.DataArray]:
+    """
+    Find the dimension by attribute name and value.
+
+    :param dataset: The dataset to search.
+    :param attr_name: The attribute name to search for.
+    :param attr_value: The attribute value to match.
+
+    :return: The matching DataArray if found, otherwise None.
+    """
+    for var in dataset.variables.values():
+        if attr_name in var.attrs and var.attrs[attr_name] == attr_value:
+            return var
+    return None
+
+
+def _get_dim_names(
+    dataset: xr.Dataset,
+) -> Tuple[Optional[xr.DataArray], Optional[xr.DataArray], Optional[xr.DataArray]]:
+    """
+    Get the latitude, longitude, and time dimensions from the dataset.
+
+    :param dataset: The dataset to search.
+
+    :return: A tuple containing (latitude, longitude, time) DataArrays.
+    """
+    lat = _get_dimension_by_attr(dataset, "units", "degrees_north")
+    lon = _get_dimension_by_attr(dataset, "units", "degrees_east")
+    time = _get_dimension_by_attr(dataset, "standard_name", "time")
+    return lat, lon, time
+
+
+def _normalise_data(arr: np.ndarray, norm_min: float, norm_max: float) -> np.ndarray:
+    if norm_min == norm_max:
+        return np.zeros_like(arr)
+    return (arr - norm_min) / (norm_max - norm_min)
+
+
+def _convert_data(
+    input_data: xr.Dataset, variable: str
+) -> Tuple[np.ndarray, float, float, str, int, xr.DataArray]:
+    """
+    Convert data to numpy array.
+
+    :param input_data: The input dataset.
+    :param variable: The variable to convert.
+
+    :return: A tuple containing the RGB data array, min value, max value, units, number
+    of longitudes, and time vector.
+    """
+    lat, lon, time = _get_dim_names(input_data)
+    num_lat, num_lon, num_time = len(lat), len(lon), len(time)
+
+    if variable == "winds_10m":
+        rgb_data_array = np.zeros((num_lat, num_lon * num_time, 3), dtype=np.uint8)
+        units = input_data["u10"].attrs.get("units", "Unknown")
+    else:
+        rgb_data_array = np.zeros((num_lat, num_lon * num_time, 1), dtype=np.uint8)
+        units = input_data[variable].attrs.get("units", "Unknown") + " * 1e-9"
+
+    variable_ranges = {
```

Comment:
Could be a constant variable at the top of the file? Out of interest, where did these values come from? What happens if we receive values outside the range?

Reply:
Moved to the top as suggested. Ranges are based on expected absolute values in the data, so I made them up ... values are always capped to this range (`np.clip`). larger ranges would reduce the contour resolution. so this is a trade off and might need to be changed once we implement the maps in the UI